src/common/src/array/stream_chunk_iter.rs

```diff
@@ -12,140 +12,177 @@
 
-use super::column::Column;
-use super::Row;
+use std::iter::once;
+
+use auto_enums::auto_enum;
+
+use super::RowRef;
 use crate::array::{Op, StreamChunk};
-use crate::types::{DatumRef, ToOwnedDatum};
 
 impl StreamChunk {
-    pub fn rows(&self) -> impl Iterator<Item = RowRef<'_>> {
+    /// Return an iterator on stream records of this stream chunk.
+    pub fn records(&self) -> impl Iterator<Item = RecordRef<'_>> {
         StreamChunkRefIter {
             chunk: self,
-            idx: 0,
+            inner: self.data_chunk.rows(),
         }
     }
+
+    /// Return an iterator on rows of this stream chunk.
+    ///
+    /// Should consider using [`StreamChunk::records`] if possible.
+    pub fn rows(&self) -> impl Iterator<Item = (Op, RowRef<'_>)> {
+        self.data_chunk.rows().map(|row| {
+            (
+                // SAFETY: index is checked since we are in the iterator.
+                unsafe { *self.ops().get_unchecked(row.index()) },
+                row,
+            )
+        })
+    }
 }
 
+type RowRefIter<'a> = impl Iterator<Item = RowRef<'a>>;
+
 struct StreamChunkRefIter<'a> {
     chunk: &'a StreamChunk,
-    idx: usize,
+
+    inner: RowRefIter<'a>,
 }
 
 impl<'a> Iterator for StreamChunkRefIter<'a> {
-    type Item = RowRef<'a>;
+    type Item = RecordRef<'a>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        match self.chunk.visibility() {
-            Some(bitmap) => {
-                loop {
-                    let idx = self.idx;
-                    if idx >= self.chunk.capacity() {
-                        return None;
-                    }
-                    // SAFETY: idx is checked.
-                    let vis = unsafe { bitmap.is_set_unchecked(idx) };
-                    self.idx += 1;
-                    if vis {
-                        return Some(RowRef {
-                            chunk: self.chunk,
-                            idx,
-                        });
-                    }
-                }
-            }
-            None => {
-                let idx = self.idx;
-                if idx >= self.chunk.capacity() {
-                    return None;
-                }
-                self.idx += 1;
-                Some(RowRef {
-                    chunk: self.chunk,
-                    idx,
+        let row = self.inner.next()?;
+        // SAFETY: index is checked since `row` is `Some`.
+        let op = unsafe { self.chunk.ops().get_unchecked(row.index()) };
+
+        match op {
+            Op::Insert => Some(RecordRef::Insert(row)),
+            Op::Delete => Some(RecordRef::Delete(row)),
+            Op::UpdateDelete => {
+                let insert_row = self.inner.next().expect("expect a row after U-");
+                // SAFETY: index is checked since `insert_row` is `Some`.
+                let op = unsafe { *self.chunk.ops().get_unchecked(insert_row.index()) };
+                debug_assert_eq!(op, Op::UpdateInsert, "expect a U+ after U-");
```

Comment:
```suggestion
                debug_assert_eq!(op, Op::UpdateInsert, "expect an U+ after U-");
```

Reply:
Should be "a" here since we pronounce it as `[juː]`.